{% extends "base.html" %}
{% load static i18n %}
{% block title %}DNase I Hypersensitive Sites{% endblock %}

{% block css %}
{{ block.super }}
<link rel="stylesheet" href="{% static 'search/css/base.css' %}">
{% endblock %}
{% block javascript %}
{{ block.super }}
<script type="text/javascript" src="{% static 'js/lodash.min.js' %}"></script>
<script type="text/javascript" src="{% static 'genoverse/js/genoverse.min.js' %}"></script>
<script defer type="text/javascript" src="{% static 'search/js/index.js' %}"></script>
{% endblock %}

{% block content %}
<div class="f1 b">DNase I Hypersensitive Sites</div>
<div>
    <div>
        <div class="pv2" id="genoverse"></div>
        <div class="f2 b">Sites</div>
        <table id="dhs" class="bg-black-10 collapse ba pv2">
            <tr class="striped--light-gray"><th class="pv2 ph3">ID</th><th class="pv2 ph3">Cell Line</th><th class="pv2 ph3">Location</th><th class="pv2 ph3">Strand</th><th class="pv2 ph3">Closest Gene</th><th class="pv2 ph3">Reference Genome</th></tr>
            {% for dhs, closest_gene in dhss %}
            <tr class="striped--light-gray">
                <td class="pv2 ph3"><a href="{% url 'search:dhs' dhs.id %}">DHS:{{ dhs.id }}</a></td>
                <td class="pv2 ph3">{{ dhs.cell_line }}</td>
                <td class="pv2 ph3"><span>{{ dhs.chromosome_name }}</span>: <span>{{ dhs.location.lower }}-{{ dhs.location.upper }}</span></td>
                <td class="pv2 ph3">{{ dhs.strand }}</td>
                <td class="pv2 ph3"><a href="{% url 'search:genes' 'ensembl' closest_gene.ensembl_id %}">{{ closest_gene.ensembl_id }} ({{ dhs.closest_gene_name }})</a></td>
                <td class="pv2 ph3">{{ dhs.ref_genome }}.{{ dhs.ref_genome_patch|default:"0" }}</td>
            </tr>
            {% endfor %}
        </table>
    </div>
</div>
{% endblock %}

{% block inline_javascript %}
<script>
    window.addEventListener("hashchange", _.throttle(function() {
        let hash = window.location.hash;
        [_, chromo, start, end] = hash.match(/^#r\=(\d+):(\d+)-(\d+)$/, '')
        fetch(`http://127.0.0.1:8000/search/dhsloc/chr${chromo}/${start}/${end}?search_type=overlap&accept=application/json&extended=t`)
            .then(response => {
                return response.json()
            })
            .then(data => {
                newTable = e("table", {id: "dhs",  class: "bg-black-10 collapse ba"}, [
                    e("tr", {class: "striped--light-gray"}, [
                        e("th", {class: "pv2 ph3"}, "ID"),
                        e("th", {class: "pv2 ph3"}, "Cell Line"),
                        e("th", {class: "pv2 ph3"}, "Location"),
                        e("th", {class: "pv2 ph3"}, "Strand"),
                        e("th", {class: "pv2 ph3"}, "Closest Gene"),
                        e("th", {class: "pv2 ph3"}, "Reference Genome"),
                    ])
                ])
                for(dhs of data) {
                    newTable.append(
                        e("tr",  {class: "striped--light-gray"},[
                            e("td", {class: "pv2 ph3"}, e("a", {href: `/search/dhs/${dhs.id}`}, `DHS:${dhs.id}`)),
                            e("td", {class: "pv2 ph3"}, dhs.cell_line),
                            e("td", {class: "pv2 ph3"}, [
                                e("span", dhs.chr),
                                ": ",
                                e("span", `${dhs.start}-${dhs.end}`)
                            ]),
                            e("td", {class: "pv2 ph3"}, dhs.strand || "None"),
                            e("td", {class: "pv2 ph3"}, e("a", {href: `/search/gene/ensembl/${dhs.closest_gene.ensembl_id}`}, `${dhs.closest_gene.ensembl_id} (${dhs.closest_gene_name})`)),
                            e("td", {class: "pv2 ph3"}, `${dhs.ref_genome}.${dhs.ref_genome_patch || 0}`),
                        ])
                    )
                }
                let dhsTable = document.getElementById("dhs");
                dhsTable.replaceWith(newTable)
            });

    }, 500, { 'leading': true }))
</script>
<script>
    new Genoverse({
    container : '#genoverse', // Where to inject Genoverse (css/jQuery selector/DOM element)
    // If no genome supplied, it must have at least chromosomeSize, e.g.:
    // chromosomeSize : 249250621, // chromosome 1, human
    genome    : 'GRCh37', // see js/genomes/
    chr       : "{{ loc.chr }}".replace("chr", ""),
    start     : {{ loc.start }},
    end       : {{ loc.end }},
    plugins   : [ 'controlPanel', [ 'karyotype', { showAssembly: true }], 'trackControls', 'resizer', 'focusRegion', 'fullscreen', 'tooltips', 'fileDrop' ],
    useHash   : true,
    hideEmptyTracks: false,
    tracks    : [
        Genoverse.Track.Scaleline.extend({ strand: false }),
        Genoverse.Track.Scalebar,
        Genoverse.Track.extend({
        name            : 'DNaseI Hypersensitive Sites',
        url             : '//127.0.0.1:8000/search/dhsloc/__CHR__/__START__/__END__?search_type=overlap&accept=application/json&format=genoverse',
        resizable       : 'auto',
        model           : Genoverse.Track.Model.extend({ dataRequestLimit : 5000000 }),
        setFeatureColor : function (f) { f.color = '#F00'; }
        }),
        Genoverse.Track.Gene,
    ]
    });
</script>
{% endblock %}
