{% extends "base.html" %}
{% load static i18n %}
{% block title %}Search Results{% endblock %}
{% block css %}
{{ block.super }}
<link href="{% static 'genoverse/css/tooltips.css' %}" rel="stylesheet" ></link>
<style>
g.tick > text {
    font-size: 2em;
}

svg > g > text {
    font-size: 2em;
}
</style>
{% endblock %}
{% block javascript %}
{{ block.super }}
<script type="text/javascript" src="{% static 'js/lodash.min.js' %}"></script>
<script type="text/javascript" src="{% static 'genoverse/js/genoverse.min.js' %}"></script>
<script defer type="text/javascript" src="{% static 'search/js/index.js' %}"></script>
<script type="text/javascript" src="{% static 'search/js/genoverse.js' %}"></script>
<script type="text/javascript" src="https://d3js.org/d3.v7.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/@observablehq/plot@0.2"></script>
{% endblock %}
{% block content %}
    <div class="flex">
        <div class="text-2xl m-2 font-bold self-end align-baseline">Results</div>
        <div class="flex-grow"></div>
        <form action="{% url 'search:results' %}" method="get" class="search">
            {{ form }}
            <input type="submit" value="Search Again" class="search-button self-end">
        </form>
    </div>
    <div id="loc-header" class="text-2xl font-bold my-4">{{ loc_search.location.chromo }}:{{ loc_search.location.range.lower }}-{{ loc_search.location.range.upper }}</div>
    {% if loc_search.features|length > 0 %}
    <div class="p-2" id="genoverse"></div>
    <div id="plot"></div>
    <div class="my-4">
        <div class="text-xl">Genes</div>
        <div>
            <table class="data-table2">
                <tr class="th-bg-light-gray"><th>ENSEMBL ID</th><th>Type</th><th>Name</th><th>Location</th><th>Strand</th><th>Reference Genome</th><th>Reg. Effect Target</th></tr>
                {% for feature, assemblies in loc_search.features.items %}
                    {% cycle 'bg-gray-200' 'th-bg-light-gray' as rowcolor silent %}
                    {% for assembly in assemblies %}
                        <tr class="{{ rowcolor }}">
                            {% if forloop.first %}
                            <th rowspan="{{ assemblies|length }}" scope="rowgroup"><a href="{% url 'search:features' 'ensembl' feature.ensembl_id %}">{{ feature.ensembl_id }}</a></th>
                            <th rowspan="{{ assemblies|length }}" scope="rowgroup">{{ feature.feature_subtype }}</th>
                            {% endif %}
                            <td>{{ assembly.name }}</td>
                            <td><span>{{ assembly.chrom_name }}</span>: <span>{{ assembly.location.lower }}-{{ assembly.location.upper|add:"-1" }}</span></td>
                            <td>{{ assembly.strand }}</td>
                            <td>{{ assembly.ref_genome }}.{{ assembly.ref_genome_patch|default:"0" }}</td>
                            <td>{% if assembly.regulatory_effects|length > 0 %}Yes{% else %}No{% endif %}</td>
                        </tr>
                    {% endfor %}
                {% endfor %}
            </table>
        </div>
    </div>
    {% endif %}
    {% if dhss.all|length > 0 %}
    <div class="my-4">
        <div class="text-xl font-bold">DNase I Hypersensitive Sites</div>
        {% include "search/partials/_dnaregion.html" with regions=dhss %}
    </div>
    {% endif %}
{% endblock %}
{% block inline_javascript %}
    {% if loc_search.features|length > 0 %}
    <script>
        // Update the chromosome location at the top of the page
        const updateLocation = function(state, key) {
            if (key !== "location") { return; }

            let location = state[key];

            const locHeaderNode = document.getElementById("loc-header");

            locHeaderNode.firstChild.data = `chr${location.chr}: ${location.start}-${location.end}`;
        }

        // Show the plot
        function geneDHSPlot(data) {
            let dhssWithTargets = data.filter(dhs => dhs.effects.length > 0 && dhs.effects.filter(effect => effect.targets.length > 0).length > 0);
            let dwtSet = new Set(dhssWithTargets);
            let targets = dhssWithTargets.flatMap(dhs => dhs.effects.flatMap(effect => effect.target_assemblies.map(assembly => {
                assembly["effect_size"] = effect.effect_size;
                assembly["label"] = dhs.label;
                return assembly
            })))

            let plot = Plot.plot({
                height: 480,
                padding: 0,
                marginLeft: 50,
                marginTop: 50,
                color: {
                    range: ["#f0e442", "#56b4e9"],
                    interpolate: "hcl"
                },
                x: {
                    axis: "top",
                    round: false,
                    label: "DHS"
                },
                y: {
                    round: false,
                    label: "Gene",
                    tickPadding: 6,
                    tickRotate: -90,
                },
                marks: [
                    Plot.cell(
                        targets,
                        {
                            x: "label",
                            y: "name",
                            fill: "effect_size",
                            title: "effect_size",
                            rx: 8,
                            ry: 8
                        }
                    ),
                    Plot.text(
                        targets,
                        {
                            x: "label",
                            y: "name",
                            text: d => d.effect_size?.toFixed(3),
                            title: "effect_size",
                        }
                    )
                ]
            })

            return plot;
        }
        function applyPlot(state, key) {
            if (key !== "dhs-effect-data" && key !== "location") { return; }
            let dhss = state["dhs-effect-data"];
            let location = state["location"];

            if (!dhss || !location) { return; }

            data = location ? dhss.filter(dhs => location.chr === dhs.chr && dhs.start >= location.start && dhs.end <= location.end) : dhss;

            let genePlot = document.getElementById("plot");
            if (genePlot) {
                let plotNode = geneDHSPlot(data)
                if (genePlot.firstChild) {
                    genePlot.replaceChild(plotNode, genePlot.firstChild);
                } else {
                    genePlot.appendChild(plotNode);
                }
            }
        }

        // Show the browser
        let browser = new CEGSGenoverse({
            container : '#genoverse', // Where to inject Genoverse (css/jQuery selector/DOM element)
            // If no genome supplied, it must have at least chromosomeSize, e.g.:
            // chromosomeSize : 249250621, // chromosome 1, human
            genome    : '{{ loc_search.assembly }}', // see js/genomes/
            assembly  : '{{ loc_search.assembly }}',
            chr       : "{{ loc_search.location.chromo }}".replace("chr", ""),
            start     : {{ loc_search.location.range.lower }},
            end       : {{ loc_search.location.range.upper }},
            plugins   : [[ 'karyotype', { showAssembly: true }], 'trackControls', 'tooltips' ],
            useHash   : true,
            hideEmptyTracks: false,
            sharedStateCallbacks: [updateLocation, applyPlot],
            width: 1200,
            tracks    : [
                Genoverse.Track.Scaleline.extend({ strand: false }),
                Genoverse.Track.Scalebar,
                Genoverse.Track.DHS.Effects.extend({children: [Genoverse.Track.Gene.extend({ legend: false })]}),
            ]
        });
    </script>
    {% endif %}
{% endblock %}
