
<!-- Modal toggle -->
<button data-modal-target="default-modal" data-modal-toggle="default-modal" class="block mt-4 lg:inline-block lg:mt-0 text-white bg-inherit hover:text-white px-2 py-2 rounded hover:bg-blue-700 no-underline" type="button">
    <i class="fa-solid fa-book"></i> Glossary
</button>
  <!-- Main modal -->
  <div id="default-modal" tabindex="-1" aria-hidden="true" class="hidden overflow-y-auto overflow-x-hidden fixed top-0 right-0 left-0 z-50 justify-center items-center w-full md:inset-0 h-[calc(100%-1rem)] max-h-full">
    <div class="relative p-4 w-full max-w-7xl max-h-full">
        <!-- Modal content -->
        <div class="relative bg-white rounded-lg shadow">
            <!-- Modal header -->
            <div class="flex items-center justify-between p-4 md:p-5 border-b rounded-t">
                <h3 class="text-xl font-semibold text-gray-900">
                    <i class="fa-solid fa-book"></i> CCGR Glossary
                </h3>
                <button type="button" class="text-gray-400 bg-transparent hover:bg-gray-200 hover:text-gray-900 rounded-lg text-sm w-8 h-8 ms-auto inline-flex justify-center items-center" data-modal-hide="default-modal">
                    <svg class="w-3 h-3" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 14 14">
                        <path stroke="currentColor" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="m1 1 6 6m0 0 6 6M7 7l6-6M7 7l-6 6"/>
                    </svg>
                    <span class="sr-only">Close modal</span>
                </button>
            </div>
            <!-- Modal body -->
            <div class="p-4 md:p-5 space-y-4">
                <p class="text-base leading-relaxed text-gray-600">
                    Duke Center for Combinatorial Gene Regulation (CCGR): The ambitious goal of the Duke Center for Combinatorial Gene Regulation, a Center of Excellence in Genomic Science, is to make combinatorial studies of the function of regulatory elements and noncoding variants routine.
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    CEGS: The Centers of Excellence in Genomic Science (CEGS) program supports multi-investigator, interdisciplinary research teams working together to address biomedical problems through development of integrated, transformative genomic approaches.
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Regulatory Elements
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    DNase I hypersensitive site (DHS): a region of chromatin in the genome that is more sensitive to cleavage by the DNase I enzyme. These regions are characterized by increased DNA accessibility, which is due to the local remodeling of the chromatin structure.<a href="#ref29">[29]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Guide RNA (gRNA): a short RNA sequence that guides the editing of RNA transcripts or DNA.<a href="#ref30">[30]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Chromatin Accessible Region (CAR): genomic regions where chromatin is open or loosened, allowing DNA-binding proteins and transcription factors to bind to their target sequences.<a href="#ref31">[31]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Called Regulatory Element(cre): Regulatory element found but label/name not assigned
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    CCRES
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    CCRE: regions of non-coding DNA which regulate the transcription of neighboring genes.<a href="#ref27">[27]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    A promoter-like sequence (PLS): A PLS is a candidate cis-regulatory element(cCRE) that's putatively labeled as a promoter.<a href="#ref27">[27]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Proximal enhancer-like (pELS): is a candidate cis-regulatory element (cCRE) that is classified as an enhancer and is located within 2 kilobases (kb) of an annotated GENCODE transcription start site (TSS).<a href="#ref27">[27]</a>                                         </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Distal enhancer-like (dELS): elements are a type of candidate cis-regulatory element (cCRE) that are classified as putative enhancers.<a href="#ref27">[27]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    H3K4me3: When combined, DNase-H3K4me3 elements are regions that exhibit both DNase hypersensitivity and H3K4me3 marks. These regions typically have promoter-like biochemical signatures and are indicative of active gene regulation.<a href="#ref28">[28]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Distal enhancer-like (dELS): elements are a type of candidate cis-regulatory element (cCRE) that are classified as putative enhancers.<a href="#ref27">[27]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Techniques and Screens
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    ATAC-STARR-seq: A massively parallel reporter assay (MPRA) that quantifies regulatory activity within accessible chromatin. It combines ATAC-seq (assay for transposase-accessible chromatin using sequencing) with STARR-seq (self-transcribing active regulatory region sequencing) to selectively assay the regulatory potential of accessible DNA.<a href="#ref1">[1]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Flow-FISH CRISPR screen: A method combining Hybridization Chain Reaction Fluorescent In-Situ Hybridization (HCR-FlowFISH) with Flow Cytometry to quantify native transcripts within their endogenous loci following CRISPR perturbations of regulatory elements.<a href="#ref2">[2]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Perturb-Seq: A high-throughput method that combines CRISPR-mediated gene inactivations with single-cell RNA sequencing (scRNA-seq) to assess comprehensive gene expression phenotypes for each perturbation.<a href="#ref3">[3]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Proliferation screen: A type of assay used to measure the rapid increase in cell numbers, often used to assess cell growth, division, and the effects of various treatments on cell proliferation.<a href="#ref4">[4]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    scCERES: single-cell (CERES) a computational method to estimate gene dependency levels from CRISPR-Cas9 essentiality screens while accounting for the copy-number-specific effect.<a href="#ref5">[5]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    wgCERES: whole-genome CERES (wgCERES) a computational method to estimate gene dependency levels from CRISPR-Cas9 essentiality screens while accounting for the copy-number-specific effect.<a href="#ref5">[5]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Cell Lines
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    iPSC (Induced Pluripotent Stem Cells): Pluripotent stem cells generated directly from a somatic cell, capable of differentiating into any cell type in the body.<a href="#ref6">[6]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    K562: A human immortalized myelogenous leukemia cell line, derived from a 53-year-old female chronic myelogenous leukemia patient.<a href="#ref7">[7]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    NPC (Neural Progenitor Cells): Cells that can differentiate into various types of neural cells, including neurons and glial cells.<a href="#ref8">[8]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    CD8: A transmembrane glycoprotein that serves as a co-receptor for the T-cell receptor (TCR), predominantly expressed on cytotoxic T cells.<a href="#ref9">[9]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    MCF-7: A breast cancer cell line isolated from a 69-year-old woman, used extensively in breast cancer research.<a href="#ref10">[10]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    HCT116: A human colon cancer cell line used in therapeutic research and drug screenings.<a href="#ref11">[11]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    DU 145: A human prostate cancer cell line used in cancer research.<a href="#ref12">[12]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    A549: A human lung carcinoma cell line used in cancer research and drug testing.<a href="#ref13">[13]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    NCI-H460: A human large cell lung cancer cell line used in cancer research.<a href="#ref14">[14]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    PC-3: A human prostate cancer cell line used in cancer research.<a href="#ref15">[15]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    MDA-MB-231: A human breast cancer cell line used in cancer research.<a href="#ref16">[16]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    HepG2: A human liver cancer cell line used in cancer research.<a href="#ref17">[17]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    SW620: A human colorectal adenocarcinoma cell line used in cancer research.<a href="#ref18">[18]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Functional Characterization Modalities
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    scCRISPRi: Single-cell CRISPR interference, a technique used to study gene function at single-cell resolution by inhibiting gene expression.<a href="#ref19">[19]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    CRISPRi: CRISPR interference, a technique that uses a catalytically dead Cas9 (dCas9) to inhibit gene expression.<a href="#ref19">[19]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    CRISPRa: CRISPR activation, a technique that uses dCas9 fused to transcriptional activators to increase gene expression.<a href="#ref20">[20]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Reporter Assay: An assay that uses a reporter gene to measure the activity of a particular gene or regulatory element.<a href="#ref21">[21]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Genome Assembly
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    hg-19: GRCh37 (hg19) an earlier human reference genome version. This version has been widely used in genomics projects, including long-term, global studies like the 1,000 Genomes Project.<a href="#ref22">[22]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    hg-38: genome assembly GRCh38 (hg38) from Genome Reference Consortium. Corrected and improved version of hg19.<a href="#ref22">[22]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Genomic Terms
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Gene: A gene is the basic physical and functional unit of heredity. Genes are made up of DNA and each chromosome contains many genes.<a href="#ref23">[23]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Transcript: A messenger RNA (mRNA) molecule that carries genetic information from DNA to the cytoplasm, where proteins are made.<a href="#ref24">[24]</a>
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Exon: The sections of DNA (or RNA) that code for proteins.<a href="#ref25">[25]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    Identifiers
                </p>
                <p class="text-base leading-relaxed text-gray-600">
                    Ensembl ID: is an identifier used by the Ensembl database to label features like genes, transcripts, exons, or proteins.<a href="#ref26">[26]</a>
                </p>
                <p class="text-base text-xl leading-relaxed text-gray-800">
                    References
                </p>
                    <ul>
                        <li id="ref1">1. <a href="https://github.com/HodgesGenomicsLab/ATAC-STARR-seq">https://github.com/HodgesGenomicsLab/ATAC-STARR-seq</a></li>
                        <li id="ref2">2. <a href="https://www.biorxiv.org/content/10.1101/2020.05.11.078675v1.full.pdf">https://www.biorxiv.org/content/10.1101/2020.05.11.078675v1.full.pdf</a></li>
                        <li id="ref3">3. <a href="https://en.wikipedia.org/wiki/Perturb-seq">https://en.wikipedia.org/wiki/Perturb-seq</a></li>
                        <li id="ref4">4. <a href="https://www.news-medical.net/life-sciences/What-is-a-Cell-Proliferation-Assay.aspx">https://www.news-medical.net/life-sciences/What-is-a-Cell-Proliferation-Assay.aspx</a></li>
                        <li id="ref5">5. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC5709193/">https://pmc.ncbi.nlm.nih.gov/articles/PMC5709193/</a></li>
                        <li id="ref6">6. <a href="https://en.wikipedia.org/wiki/Induced_pluripotent_stem_cell">https://en.wikipedia.org/wiki/Induced_pluripotent_stem_cell</a></li>
                        <li id="ref7">7. <a href="https://en.wikipedia.org/wiki/K562_cells">https://en.wikipedia.org/wiki/K562_cells</a></li>
                        <li id="ref8">8. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC6291443/">https://pmc.ncbi.nlm.nih.gov/articles/PMC6291443/</a></li>
                        <li id="ref9">9. <a href="https://en.wikipedia.org/wiki/CD8">https://en.wikipedia.org/wiki/CD8</a></li>
                        <li id="ref10">10. <a href="https://en.wikipedia.org/wiki/MCF-7">https://en.wikipedia.org/wiki/MCF-7</a></li>
                        <li id="ref11">11. <a href="https://en.wikipedia.org/wiki/HCT116_cells">https://en.wikipedia.org/wiki/HCT116_cells</a></li>
                        <li id="ref12">12. <a href="https://en.wikipedia.org/wiki/DU145">https://en.wikipedia.org/wiki/DU145</a></li>
                        <li id="ref13">13. <a href="https://en.wikipedia.org/wiki/A549_cell">https://en.wikipedia.org/wiki/A549_cell</a></li>
                        <li id="ref14">14. <a href="https://en.wikipedia.org/wiki/NCI-60#NCI-H460">https://en.wikipedia.org/wiki/NCI-60#NCI-H460</a></li>
                        <li id="ref15">15. <a href="https://en.wikipedia.org/wiki/PC3">https://en.wikipedia.org/wiki/PC3</a></li>
                        <li id="ref16">16. <a href="https://en.wikipedia.org/wiki/MDA-MB231">https://en.wikipedia.org/wiki/MDA-MB231</a></li>
                        <li id="ref17">17. <a href="https://en.wikipedia.org/wiki/Hep_G2">https://en.wikipedia.org/wiki/Hep_G2</a></li>
                        <li id="ref18">18. <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC5783577/">https://pmc.ncbi.nlm.nih.gov/articles/PMC5783577/</a></li>
                        <li id="ref19">19. <a href="https://en.wikipedia.org/wiki/CRISPR_interference">https://en.wikipedia.org/wiki/CRISPR_interference</a></li>
                        <li id="ref20">20. <a href="https://en.wikipedia.org/wiki/CRISPR_activation">https://en.wikipedia.org/wiki/CRISPR_activation</a></li>
                        <li id="ref21">21. <a href="https://en.wikipedia.org/wiki/Reporter_gene">https://en.wikipedia.org/wiki/Reporter_gene</a></li>
                        <li id="ref22">22. <a href="https://www.ncbi.nlm.nih.gov/grc/data">https://www.ncbi.nlm.nih.gov/grc/data</a></li>
                        <li id="ref23">23. <a href="https://medlineplus.gov/genetics/understanding/basics/gene/">https://medlineplus.gov/genetics/understanding/basics/gene/</a></li>
                        <li id="ref24">24. <a href="https://www.britannica.com/science/messenger-RNA">https://www.britannica.com/science/messenger-RNA</a></li>
                        <li id="ref25">25. <a href="https://www.nature.com/scitable/definition/intron-introns-67/">https://www.nature.com/scitable/definition/intron-introns-67/</a></li>
                        <li id="ref26">26. <a href="https://mart.ensembl.org/info/genome/stable_ids/index.html">https://mart.ensembl.org/info/genome/stable_ids/index.html</a></li>
                        <li id="ref27">27. <a href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm10&g=encodeCcreCombined">https://genome.ucsc.edu/cgi-bin/hgTrackUi?db=mm10&g=encodeCcreCombined</a></li>
                        <li id="ref28">28. <a href="https://en.wikipedia.org/wiki/H3K4me3">https://en.wikipedia.org/wiki/H3K4me3</a></li>
                        <li id="ref29">29. <a href="https://en.wikipedia.org/wiki/DNase_I_hypersensitive_site">https://en.wikipedia.org/wiki/DNase_I_hypersensitive_site</a></li>
                        <li id="ref30">30. <a href="https://en.wikipedia.org/wiki/Guide_RNA">https://en.wikipedia.org/wiki/Guide_RNA</a></li>
                        <li id="ref31">31. <a href="https://en.wikipedia.org/wiki/Epigenomics#Chromatin_accessibility_assays">https://en.wikipedia.org/wiki/Epigenomics#Chromatin_accessibility_assays</a></li>
                    </ul>
            </div>
            <!-- Modal footer -->
            <div class="flex items-center p-4 md:p-5 border-t border-gray-200 rounded-b">
                <button data-modal-hide="default-modal" type="button" class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center">Exit</button>
            </div>
        </div>
    </div>
  </div>
